src/Common/Commands.Common.Authentication/Authentication/ManagedServiceAccessToken.cs

```diff
@@ -0,0 +1,132 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+using Microsoft.Azure.Commands.Common.Authentication.Properties;
+using System;
+using System.Net.Http;
+using System.Threading;
+
+namespace Microsoft.Azure.Commands.Common.Authentication
+{
+    public class ManagedServiceAccessToken : IAccessToken
+    {
+        IAzureAccount _account;
+        string _tenant;
+        string _resourceId;
+        IHttpOperations<ManagedServiceTokenInfo> _tokenGetter;
+        DateTime _expiration = DateTime.UtcNow;
+        string _accessToken;
+        string _requestUri;
+
+        public ManagedServiceAccessToken(IAzureAccount account, IAzureEnvironment environment, string resourceId, string tenant = "Common")
+        {
+            if (account == null || string.IsNullOrEmpty(account.Id) || !account.IsPropertySet(AzureAccount.Property.MSILoginUri))
+            {
+                throw new ArgumentNullException(nameof(account));
+            }
+
+            if (string.IsNullOrWhiteSpace(tenant))
+            {
+                throw new ArgumentNullException(nameof(tenant));
+            }
+            
+            if (environment == null)
+            {
+                throw new ArgumentNullException(nameof(environment));
+            }
+
+            _account = account;
+            _resourceId = GetResource(resourceId, environment);
+            var baseUri = _account.GetProperty(AzureAccount.Property.MSILoginUri);
+            var builder = new UriBuilder(baseUri);
+            builder.Query = string.Format("resource={0}", Uri.EscapeDataString(_resourceId));
+            _requestUri = builder.Uri.ToString();
+            _tenant = tenant;
+            IHttpOperationsFactory factory;
+            if (!AzureSession.Instance.TryGetComponent(HttpClientOperationsFactory.Name, out factory))
+            {
+                factory = HttpClientOperationsFactory.Create();
+            }
+
+            _tokenGetter = factory.GetHttpOperations<ManagedServiceTokenInfo>().WithHeader("Metadata", new[] { "true" });
+        }
+
+        public string AccessToken
+        {
+            get
+            {
+                try
+                {
+                    GetOrRenewAuthentication();
+                }
+                catch (HttpRequestException httpException)
+                {
+                    throw new InvalidOperationException(string.Format(Resources.MSITokenRequestFailed, _resourceId, _requestUri), httpException);
+                }
+
+                return _accessToken;
+            }
+        }
+
+        public string LoginType
+        {
+            get
+            {
+                return "ManagedService";
+            }
+        }
```

Comment:
These can all be expression-bodied members.  E.g.,

```
public string LoginType => "ManagedService"
```

Reply:
These are implementing an interface

Reply:
That will work, I think? 😄 

An expression-bodied member is equivalent to a `public get;`...?   Maybe?